pkg/prometheus/promcfg.go

```diff
@@ -2318,11 +2320,23 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 		cfg = append(cfg, yaml.MapItem{Key: "scrape_timeout", Value: *sc.Spec.ScrapeTimeout})
 	}
 
-	if sc.Spec.RelabelConfigs != nil {
+	if len(sc.Spec.RelabelConfigs) > 0 {
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
-		cfg = append(cfg, yaml.MapItem{Key: "relabel_configs", Value: relabelings})
 	}
 
+	alreadyHasShardLabel := false
+	for _, relabelConfig := range sc.Spec.RelabelConfigs {
+		if relabelConfig.Action == "hashmod" {
+			level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "relabeling", relabelConfig)
+			alreadyHasShardLabel = true
+			break
+		}
+	}
+	if !alreadyHasShardLabel {
+		relabelings = generateAddressShardingRelabelingRules(relabelings, shards)
+	}
```

Comment:
I think we want to wrap this inside a conditional `if shards > 1`

Seems wrong to add shard-related relabelings even though Prometheus isn't sharded 🤔

Reply:
The initial reasoning was probably that the extra branch makes the code a bit less readable for little gain.

Reply:
I imagine that Prometheus will need to perform extra CPU work for nothing if we add a no-op relabeling rule. I don't know exactly how much of an impact this could have, but I can imagine a few CPU cycles going to waste for very large-scale Prometheus :P

Reply:
I've reverted the change and the skipping non-sharded configs will be done again now. I agree on not adding a relabelling rule, might be confusing for users that do not shard as well.